File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cc
#include <iostream>
#include <algorithm>
using namespace std;
using ll = long long;

ll tree[1'000'000 * 4 + 16];

void update(ll left, ll right, ll idx, ll cvalue, ll cidx) {
	if (cidx < left || right < cidx) { //범위밖인경우 -> 암것도안하고리턴
		return;
	}
	if (left == right) { //리프노드인경우 = 트리최하단도달 -> 값저장
		tree[idx] = cvalue;
		return;
	}
	//분할
	ll mid = (left + right) / 2;
	update(left, mid, idx * 2, cvalue, cidx);
	update(mid+1, right, idx * 2 + 1, cvalue, cidx);
	tree[idx] = tree[idx * 2] + tree[idx * 2 + 1];
	return;
}

ll query(ll left, ll right, ll idx, ll beg, ll end) { //left,right = 노드의담당범위 / beg,end = 구하고자하는범위
	if (end < left || right < beg) { //범위밖인경우 -> 0리턴
		return 0;
	}
	if (beg <= left && right <= end) { //left~right가 beg~end의 부분집합인경우 통째로 리턴
		return tree[idx];
	}
	//분할
	ll mid = (left + right) / 2;
	return query(left, mid, idx * 2, beg, end) + query(mid + 1, right, idx * 2 + 1, beg, end);
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);


	int n, m, k;
	cin >> n >> m >> k;
	for (int i = 1; i <= n; i++) {
		ll x;
		cin >> x;
		update(1, n, 1, x, i);
	}
	for (int i = 0; i < m + k; i++) {
		ll a, b, c;
		cin >> a >> b >> c;
		if (a == 1) {
			update(1, n, 1, c, b);
		}
		else {
			cout << 1LL * query(1, n, 1, b, c) << "\n";
		}
	}
	return 0;
}

File: 백준/Gold/2042. 구간 합 구하기/구간 합 구하기.cs
using System;
using System.Runtime.InteropServices;
using ll = System.Int64;

namespace CSharpTest
{
    internal class Program
    {

        static ll[] tree = new ll[1000000 * 4 + 16];

        static void Update(ll idx, ll left, ll right, ll cidx, ll cval)
        {
            if (cidx < left || right < cidx)
            { 
                return;
            }
            if (left == right)
            {
                tree[idx] = cval;
                return;
            }
            ll mid = (left + right) / 2;
            Update(idx * 2, left, mid, cidx, cval);
            Update(idx * 2 + 1, mid + 1, right, cidx, cval);
            tree[idx] = tree[idx * 2] + tree[idx * 2 + 1];
            return;
        }
        static ll Query(ll idx, ll left, ll right, ll qleft, ll qright)
        {
            if (qright < left || right < qleft)
            {
                return 0;
            }
            if (qleft <= left && right <= qright)
            {
                return tree[idx];
            }
            ll mid = (left + right) / 2;
            return Query(idx * 2, left, mid, qleft, qright) + Query(idx * 2 + 1, mid + 1, right, qleft, qright);
        }

        static void Main(string[] args)
        {
            ll n, m, k;
            string[] buf1 = Console.ReadLine().Split();
            n = ll.Parse(buf1[0]);
            m = ll.Parse(buf1[1]);
            k = ll.Parse(buf1[2]);

            for (int i = 1; i <= n; i++)
            {
                ll x = ll.Parse(Console.ReadLine());
                Update(1, 1, n, i, x);
            }

            for(int i= 1; i <= m+k; i++)
            {
                string[] buf2 = Console.ReadLine().Split();
                ll a = ll.Parse(buf2[0]);
                ll b = ll.Parse(buf2[1]);
                ll c = ll.Parse(buf2[2]);
                if (a == 1)
                {
                    Update(1, 1, n, b, c);
                } else
                {
                    Console.WriteLine(Query(1,1,n,b,c));
                }
            }

        }
    }
}
